src/mlpack/tests/neat_test_tools.hpp

```diff
@@ -0,0 +1,296 @@
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/continuous_multiple_pole_cart.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/multiple_pole_cart.hpp>
+#include <mlpack/methods/ann/activation_functions/hard_sigmoid_function.hpp>
+#include <mlpack/methods/neat/neat.hpp>
+
+using namespace mlpack;
+using namespace mlpack::neat;
+using namespace mlpack::ann;
+using namespace mlpack::rl;
+
+// A class that defines the XOR task.
+class XORTask
+{
+ public:
+  double Evaluate(Genome<HardSigmoidFunction>& genome)
+  {
+    arma::mat input = {{0, 0, 1, 1},
+                       {0, 1, 0, 1}};
+
+    double error = 0;
+    for (size_t i = 0; i < input.n_cols; i++)
+    {
+      arma::vec inputVec = input.col(i);
+      arma::vec output = genome.Evaluate(inputVec);
+      arma::vec answer = {(input(0, i) + input(1, i)) * (!input(0, i) +
+          !input(1, i))};
+      error += std::pow(answer[0] - output[0], 2);
+    }
+    return 4 - error;
+  }
+
+  template <typename Archive>
+  void serialize(Archive& /* ar */, const unsigned int /* version */)
+  { /* Nothing to do here */}
+};
+
+/**
+ * A Task class that wraps a continuous RL environment.
+ */
+template<class EnvironmentType>
+class ContinuousRLTask
+{
+ public:
+  ContinuousRLTask(const EnvironmentType& env) : environment(env)
+  { /* Nothing to do here */ }
+
+  double Evaluate(Genome<HardSigmoidFunction>& genome)
+  {
+
+    double total = 0;
+    for (size_t i = 0; i < 5; i++)
+    {
+      // Set the initial state.
+      typename EnvironmentType::State state = environment.InitialSample();
+      arma::vec output = genome.Evaluate(state.Data());
+      double fitness = 0;
+      size_t k = 0;
+      while (!environment.IsTerminal(state))
+      {
+        // std::cout << "Timestep: " << k << std::endl;
+        k++;
+        // Choose the action to perform.
+        typename EnvironmentType::Action action;
+        action.action[0] = output[0];
+
+        // Use the current action to get the next state.
+        // std::cout << "Getting next state" << std::endl;
+        fitness += environment.Sample(state, action, state);
+
+        // Update the state of the genome for the next step.
+        // std::cout << "Finding output" << std::endl;
+        output = genome.Evaluate(state.Data());
+
+        if (k >= 10000)
+          break;
+      }
+      fitness += 500;
+      total += fitness;
+    }
+
+    std::cout << "Fitness: " << total / 5 << std::endl;
+    return total / 5;
+  }
+
+ private:
+  EnvironmentType environment;
+};
+
+/**
+ * A Task class that wraps a discrete RL environment.
+ */
+template<class EnvironmentType>
+class DiscreteRLTask
+{
+ public:
+  DiscreteRLTask(const EnvironmentType& env) : environment(env)
+  { /* Nothing to do here */ }
+
+  double Evaluate(Genome<HardSigmoidFunction>& genome)
+  {
+    // Set the initial state.
+    typename EnvironmentType::State state = environment.InitialSample();
```

Comment:
We want to evaluate if a model is all to solve a task using different start conditions, so this should move down into the for loop, so that we start with a new random position instead of using the previous one.